File: 4.Counting Element - MaxCounters/answer.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")

def solution(N, A):
    # write your code in Python 3.6
    NPlusOne = N + 1
    counters = [0] * N
    maxVal = 0
    last_max = 0
    for i in A:
        if(i == NPlusOne):
            last_max = maxVal
        else:
            if(counters[i-1] + 1 > last_max):
                counters[i-1] += 1
            else:
                counters[i-1] = last_max + 1
                
            if(counters[i-1] > maxVal):
                maxVal = counters[i-1]
                
    for i in range(len(counters)):
        if(counters[i] < last_max):
            counters[i] = last_max
    return counters

File: 4.Counting Element - MaxCounters/answer.js
// you can write to stdout for debugging purposes, e.g.
// console.log('this is a debug message');

function solution(N, A) {
    // write your code in JavaScript (Node.js 8.9.4)
    let answer = Array(N).fill(0);
    let maxCounterVal = 0;
    let currentMax = 0
    A.forEach(num => {
        index = num - 1
        if(num >= N + 1){
            maxCounterVal = currentMax
        }else{
            if(maxCounterVal > answer[index])
                answer[index] = maxCounterVal + 1
            else
                answer[index] += 1
            
            if(currentMax < answer[index])
                currentMax = answer[index]
        }
    })
    
    answer.forEach((num, index) => {
        if(answer[index] < maxCounterVal)
            answer[index] = maxCounterVal
    })
    
    return answer
}


// you can write to stdout for debugging purposes, e.g.
// console.log('this is a debug message');

function solution(N, A) {
    // write your code in JavaScript (Node.js 8.9.4)
    let maxCounter = 0
    let currentMax = 0
    let ans = new Array(N).fill(0)
    for (val of A){
        if(val > N){
            maxCounter = currentMax
        }else{
            let index = val -1
            if(ans[index] < maxCounter)
                ans[index] = maxCounter
            
            ans[index] += 1
            if(ans[index] > currentMax)
                currentMax = ans[index]
        }
    }
    ans = ans.map( val => {
        return val > maxCounter ? val : maxCounter
    })
    return ans
}
